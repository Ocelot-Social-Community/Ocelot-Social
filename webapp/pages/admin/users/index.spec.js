import { mount } from '@vue/test-utils'
import Vuex from 'vuex'
import Users from './index.vue'

const localVue = global.localVue

const stubs = {
  'nuxt-link': true,
}

describe('Users', () => {
  let wrapper

  const mocks = {
    $t: jest.fn((t) => t),
    $apollo: {
      loading: false,
      mutate: jest
        .fn()
        .mockRejectedValue({ message: 'Ouch!' })
        .mockResolvedValue({
          data: {
            switchUserRole: {
              id: 'user',
              email: 'user@example.org',
              name: 'User',
              role: 'moderator',
              slug: 'user',
            },
          },
        }),
    },
    $toast: {
      error: jest.fn(),
      success: jest.fn(),
    },
  }

  const getters = {
    'auth/isAdmin': () => true,
    'auth/user': () => {
      return { id: 'admin' }
    },
  }

  const Wrapper = () => {
    const store = new Vuex.Store({ getters })
    return mount(Users, {
      mocks,
      localVue,
      store,
      stubs,
      data: () => ({
        User: [
          {
            id: 'user',
            email: 'user@example.org',
            name: 'User',
            role: 'moderator',
            slug: 'user',
          },
          {
            id: 'user2',
            email: 'user2@example.org',
            name: 'User',
            role: 'moderator',
            slug: 'user',
          },
        ],
      }),
    })
  }

  describe('given badges are enabled', () => {
    beforeEach(() => {
      mocks.$env = {
        BADGES_ENABLED: true,
      }
      wrapper = Wrapper()
    })

    it('renders', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('given badges are disabled', () => {
    beforeEach(() => {
      mocks.$env = {
        BADGES_ENABLED: false,
      }
      wrapper = Wrapper()
    })

    it('renders', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('search', () => {
    let searchAction
    beforeEach(() => {
      wrapper = Wrapper()
      searchAction = (wrapper, { query }) => {
        wrapper.find('input').setValue(query)
        wrapper.find('form').trigger('submit')
        return wrapper
      }
    })

    describe('query looks like an email address', () => {
      it('searches users for exact email address', async () => {
        const wrapper = await searchAction(Wrapper(), { query: 'email@example.org' })
        expect(wrapper.vm.email).toEqual('email@example.org')
        expect(wrapper.vm.filter).toBe(null)
      })

      it('email address is case-insensitive', async () => {
        const wrapper = await searchAction(Wrapper(), { query: 'eMaiL@example.org' })
        expect(wrapper.vm.email).toEqual('email@example.org')
        expect(wrapper.vm.filter).toBe(null)
      })
    })

    describe('query is just text', () => {
      it('tries to find matching users by `name`, `slug` or `about`', async () => {
        const wrapper = await searchAction(await Wrapper(), { query: 'Find me' })
        const expected = {
          OR: [
            { name_contains: 'Find me' },
            { slug_contains: 'Find me' },
            { about_contains: 'Find me' },
          ],
        }
        expect(wrapper.vm.email).toBe(null)
        expect(wrapper.vm.filter).toEqual(expected)
      })
    })
  })

  describe('change roles', () => {
    beforeAll(() => {
      wrapper = Wrapper()
      wrapper.setData({
        User: [
          {
            id: 'admin',
            email: 'admin@example.org',
            name: 'Admin',
            role: 'admin',
            slug: 'admin',
          },
          {
            id: 'user',
            email: 'user@example.org',
            name: 'User',
            role: 'user',
            slug: 'user',
          },
        ],
        userRoles: ['user', 'moderator', 'admin'],
      })
    })

    it('cannot change own role', () => {
      const adminRow = wrapper.findAll('tr').at(1)
      expect(adminRow.find('select').exists()).toBe(false)
    })

    it('changes the role of another user', () => {
      const userRow = wrapper.findAll('tr').at(2)
      userRow.findAll('option').at(1).setSelected()
      expect(mocks.$apollo.mutate).toHaveBeenCalledWith(
        expect.objectContaining({
          variables: {
            id: 'user',
            role: 'moderator',
          },
        }),
      )
    })

    it('toasts a success message after role has changed', () => {
      const userRow = wrapper.findAll('tr').at(2)
      userRow.findAll('option').at(1).setSelected()
      expect(mocks.$toast.success).toHaveBeenCalled()
    })
  })
})
