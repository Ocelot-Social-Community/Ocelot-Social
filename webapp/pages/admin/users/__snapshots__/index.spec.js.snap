// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users given badges are disabled renders 1`] = `
<div
  class="admin-users"
>
  <article
    class="base-card"
  >
    <h2
      class="title"
    >
      admin.users.name
    </h2>
     
    <form
      autocomplete="off"
      class="ds-form"
      novalidate="novalidate"
    >
      <div
        class="ds-flex"
        style="margin-left: -8px; margin-right: -8px;"
      >
        <div
          class="ds-flex-item"
          style="flex-basis: 90%; width: 90%; padding-left: 8px; padding-right: 8px; margin-bottom: 16px;"
        >
          <div
            class="ds-form-item ds-input-size-base"
          >
            <label
              class="ds-input-label"
              style="display: none;"
            >
              
  

            </label>
            <div
              class="ds-input-wrap"
            >
              <div
                class="ds-input-icon"
              >
                <span
                  aria-label="icon"
                  class="ds-icon"
                >
                  <svg
                    class="ds-icon-svg"
                    viewBox="0 0 32 32"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M19 3c5.511 0 10 4.489 10 10s-4.489 10-10 10a9.923 9.923 0 01-6.313-2.25l-7.969 7.969-1.438-1.438 7.969-7.969a9.919 9.919 0 01-2.25-6.313c0-5.511 4.489-10 10-10zm0 2c-4.43 0-8 3.57-8 8s3.57 8 8 8 8-3.57 8-8-3.57-8-8-8z"
                    />
                  </svg>
                </span>
              </div>
              <input
                class="ds-input ds-input-has-icon"
                name="query"
                placeholder="admin.users.form.placeholder"
                tabindex="0"
                type="text"
              />
              <!---->
            </div>
            <transition-stub
              name="ds-input-error"
            >
              <div
                class="ds-input-error"
                style="display: none;"
              >
                
    
  
              </div>
            </transition-stub>
          </div>
        </div>
         
        <div
          class="ds-flex-item"
          style="flex-basis: 30px; width: 30px; padding-left: 8px; padding-right: 8px; margin-bottom: 16px;"
        >
          <button
            class="base-button --icon-only --circle --filled"
            type="submit"
          >
            <span
              class="base-icon"
            >
              <!---->
            </span>
             
            <!---->
             
          </button>
        </div>
      </div>
    </form>
     
    <!---->
  </article>
   
  <article
    class="base-card"
  >
    <div
      class="ds-table-wrap"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="ds-table ds-table-condensed ds-table-bordered"
      >
        <colgroup>
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
        </colgroup>
        <thead>
          <tr>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.number
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.name
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.email
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.slug
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.createdAt
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          🖉
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          🗨
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          ❤
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          admin.users.table.columns.role
        
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="ds-table-col"
            >
              NaN.
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  User
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              <a
                href="mailto:user@example.org"
              >
                <b>
                  user@example.org
                </b>
              </a>
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  user
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              
        
      
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              <select />
            </td>
          </tr>
          <tr>
            <td
              class="ds-table-col"
            >
              NaN.
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  User
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              <a
                href="mailto:user2@example.org"
              >
                <b>
                  user2@example.org
                </b>
              </a>
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  user
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              
        
      
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              <select />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
     
    <div
      class="pagination-buttons"
    >
      <button
        class="previous-button base-button --icon-only --circle"
        data-test="previous-button"
        disabled="disabled"
        type="button"
      >
        <span
          class="base-icon"
        >
          <!---->
        </span>
         
        <!---->
         
      </button>
       
      <!---->
       
      <button
        class="next-button base-button --icon-only --circle"
        data-test="next-button"
        disabled="disabled"
        type="button"
      >
        <span
          class="base-icon"
        >
          <!---->
        </span>
         
        <!---->
         
      </button>
    </div>
     
    <!---->
  </article>
</div>
`;

exports[`Users given badges are enabled renders 1`] = `
<div
  class="admin-users"
>
  <article
    class="base-card"
  >
    <h2
      class="title"
    >
      admin.users.name
    </h2>
     
    <form
      autocomplete="off"
      class="ds-form"
      novalidate="novalidate"
    >
      <div
        class="ds-flex"
        style="margin-left: -8px; margin-right: -8px;"
      >
        <div
          class="ds-flex-item"
          style="flex-basis: 90%; width: 90%; padding-left: 8px; padding-right: 8px; margin-bottom: 16px;"
        >
          <div
            class="ds-form-item ds-input-size-base"
          >
            <label
              class="ds-input-label"
              style="display: none;"
            >
              
  

            </label>
            <div
              class="ds-input-wrap"
            >
              <div
                class="ds-input-icon"
              >
                <span
                  aria-label="icon"
                  class="ds-icon"
                >
                  <svg
                    class="ds-icon-svg"
                    viewBox="0 0 32 32"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M19 3c5.511 0 10 4.489 10 10s-4.489 10-10 10a9.923 9.923 0 01-6.313-2.25l-7.969 7.969-1.438-1.438 7.969-7.969a9.919 9.919 0 01-2.25-6.313c0-5.511 4.489-10 10-10zm0 2c-4.43 0-8 3.57-8 8s3.57 8 8 8 8-3.57 8-8-3.57-8-8-8z"
                    />
                  </svg>
                </span>
              </div>
              <input
                class="ds-input ds-input-has-icon"
                name="query"
                placeholder="admin.users.form.placeholder"
                tabindex="0"
                type="text"
              />
              <!---->
            </div>
            <transition-stub
              name="ds-input-error"
            >
              <div
                class="ds-input-error"
                style="display: none;"
              >
                
    
  
              </div>
            </transition-stub>
          </div>
        </div>
         
        <div
          class="ds-flex-item"
          style="flex-basis: 30px; width: 30px; padding-left: 8px; padding-right: 8px; margin-bottom: 16px;"
        >
          <button
            class="base-button --icon-only --circle --filled"
            type="submit"
          >
            <span
              class="base-icon"
            >
              <!---->
            </span>
             
            <!---->
             
          </button>
        </div>
      </div>
    </form>
     
    <!---->
  </article>
   
  <article
    class="base-card"
  >
    <div
      class="ds-table-wrap"
    >
      <table
        cellpadding="0"
        cellspacing="0"
        class="ds-table ds-table-condensed ds-table-bordered"
      >
        <colgroup>
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
          <col
            width=""
          />
        </colgroup>
        <thead>
          <tr>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.number
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.name
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.email
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.slug
        
            </th>
            <th
              class="ds-table-head-col"
            >
              
          admin.users.table.columns.createdAt
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          🖉
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          🗨
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          ❤
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          admin.users.table.columns.role
        
            </th>
            <th
              class="ds-table-head-col ds-table-head-col-right"
            >
              
          admin.users.table.columns.badges
        
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="ds-table-col"
            >
              NaN.
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  User
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              <a
                href="mailto:user@example.org"
              >
                <b>
                  user@example.org
                </b>
              </a>
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  user
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              
        
      
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              <select />
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <button
                  class="base-button --icon-only --circle --filled"
                  type="button"
                >
                  <span
                    class="base-icon"
                  >
                    <!---->
                  </span>
                   
                  <!---->
                   
                </button>
              </nuxt-link-stub>
            </td>
          </tr>
          <tr>
            <td
              class="ds-table-col"
            >
              NaN.
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  User
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              <a
                href="mailto:user2@example.org"
              >
                <b>
                  user2@example.org
                </b>
              </a>
            </td>
            <td
              class="ds-table-col"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <b>
                  user
                </b>
              </nuxt-link-stub>
            </td>
            <td
              class="ds-table-col"
            >
              
        
      
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              
            
          
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              <select />
            </td>
            <td
              class="ds-table-col ds-table-col-right"
            >
              <nuxt-link-stub
                to="[object Object]"
              >
                <button
                  class="base-button --icon-only --circle --filled"
                  type="button"
                >
                  <span
                    class="base-icon"
                  >
                    <!---->
                  </span>
                   
                  <!---->
                   
                </button>
              </nuxt-link-stub>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
     
    <div
      class="pagination-buttons"
    >
      <button
        class="previous-button base-button --icon-only --circle"
        data-test="previous-button"
        disabled="disabled"
        type="button"
      >
        <span
          class="base-icon"
        >
          <!---->
        </span>
         
        <!---->
         
      </button>
       
      <!---->
       
      <button
        class="next-button base-button --icon-only --circle"
        data-test="next-button"
        disabled="disabled"
        type="button"
      >
        <span
          class="base-icon"
        >
          <!---->
        </span>
         
        <!---->
         
      </button>
    </div>
     
    <!---->
  </article>
</div>
`;
